### ISSUE TITLE
RIN with 100k steps, still noisy images?

### ISSUE THREAD

**Issue by Samin100** [lucidrains/recurrent-interface-network-pytorch#7]
I'm trying to reproduce the [_On the Importance of Noise Scheduling for Diffusion Models_ paper's](https://arxiv.org/abs/2301.10972) benchmark on ImageNet 256 x 256 with the RIN. However, after training for around ~100k steps, the sampled images still look like noise. I've applied the fix [mentioned here](https://github.com/lucidrains/recurrent-interface-network-pytorch/issues/6) prior to training.

What I found interesting is that the noisy images change - it's almost as if the noise gets "smoother" after training, with a checkerboard pattern in between (possibly due to the image patches)?

Start (100 steps):
![media_images_samples_101_83989238bf614c33120e](https://user-images.githubusercontent.com/13106423/222613758-79db9209-b4d7-4d4a-b9fc-7a095e00fddd.png)

10k steps:
![media_images_samples_9999_dfc00accb4ad002526df](https://user-images.githubusercontent.com/13106423/222614048-a2a6f6f7-5d15-4128-b153-a0ade5e78824.png)

30k steps:
![media_images_samples_29997_36ffcfa0844d90919794](https://user-images.githubusercontent.com/13106423/222614072-55aba7f2-30ba-4352-8c36-ee8cad9ac7d6.png)

50k steps:
![media_images_samples_49995_f556fbc2b8743d825e1b](https://user-images.githubusercontent.com/13106423/222614083-1d86010a-e8b2-458e-8225-a0dfd881c743.png)

90k steps:
![media_images_samples_87062_7e6d97b7ee20825ab96b](https://user-images.githubusercontent.com/13106423/222614016-259751f3-f34f-4159-b193-4fa85f5eb26c.png)

Also, the training loss seems to be very noisy as well.
![CleanShot 2023-03-02 at 18 16 28](https://user-images.githubusercontent.com/13106423/222614630-f926899c-b48b-45b1-9572-1b5db453318c.png)

Any pointers on how to go about debugging this would be much appreciated.

**Comment by Lamikins** [comment]
What are the hyperparameters you're using? I am running into similar problems as yours on larger images: at 32x32 the model converges but higher resolutions dont converge at all. I think batch size is of critical importance to pixel-level models: in this paper they use bs=1024, and the recent simple diffusion paper uses a batch size of 2048, which the authors note is important for convergence.

**Comment by Samin100** [comment]
I’m using batch size of 256 (this is as much as can fit into memory) with a learning rate of 2e-3 with the LAMB optimizer (this is the same optimizer and learning rate used in the paper.)

@Lamikins Could you share more details around the hyperparameters that got 32 x 32 model to converge? Were you able to sample from it with good results? 

I’m trying to figure out if the model just needs  to be trained for longer or if there’s a bug in my implementation.

**Comment by lucidrains** [comment]
@Samin100 hmm, i was definitely able to make it converge for my toy oxford flowers dataset with a batch size of 32, and under 100k steps..

**Comment by lucidrains** [comment]
@Samin100 could you share your network hyperparameters?

**Comment by lucidrains** [comment]
lol, maybe this is [the issue](https://github.com/lucidrains/recurrent-interface-network-pytorch/issues/8) :face_exhaling:

**Comment by Samin100** [comment]
Yes, here's the RIN's hyperparameters. I'm mostly trying to stick to the hyperparameters recommended in the paper.

```
model = RIN(
    dim = 256,                  # model dimensions
    image_size = img_size,      # image size
    patch_size = 8,             # patch size
    depth = 4,                  # depth
    num_latents = 128,          # number of latents. they used 256 in the paper
    dim_latent = 768,           # can be greater than the image dimension (dim) for greater capacity
    latent_self_attn_depth = 6, # number of latent self attention blocks per recurrent step, K in the paper
).cuda()
```

And the diffusion wrapper:

```
diffusion = GaussianDiffusion(
    model,
    timesteps = 400,
    train_prob_self_cond = 0.9,
    scale = 1.,
    objective = 'eps',
    use_ddim=True,
    noise_schedule = 'linear',
).cuda()
```



I just started a new training run using the latest fix from #8 on 32x32 images and it's only 3k steps in, but the model still seems to only produce noise. This using batch size 1536 with lr 1e-4 and the above hyperparams.

Here's a sample from 100 steps in:
![media_images_samples_101_3f2900bee7d5098da2af](https://user-images.githubusercontent.com/13106423/222874021-8aeb3a05-0310-4d39-92e8-10f674080406.png)

3k steps:
![media_images_samples_2828_38d8624e17de0e6a06e0](https://user-images.githubusercontent.com/13106423/222874026-4702af9a-8aa4-42a0-baf3-2900861ae8be.png)

**Comment by lucidrains** [comment]
yeah that looks good, 3k steps is nothing, just keep training

**Comment by lucidrains** [comment]
@Samin100 are you seeing it now?

**Comment by lucidrains** [comment]
@Samin100 haha, i see you live in San Francisco, happy to show you in person this weekend if you can't get it working

**Comment by Samin100** [comment]
@lucidrains Gotcha! I wasn't sure if the samples were supposed to look like that at 3k steps so I stopped the training run before going to bed. I just restarted it and will share more samples later today.

I'm also doing a second training run with the new time conditioning as mentioned in #9 while keeping all other hyperparams identical, so hopefully that's a good ablation study.

 And I appreciate the help! If I can't get either of these runs to work then I might have to take you up on that offer ☕

**Comment by lucidrains** [comment]
@Samin100 ok, let me know! i was just doing a run with the latent token time conditioning, but it wasn't working so hot

i've started a run with that off (`latent_token_time_cond = False`) just to verify it can still converge with the latest fixes

**Comment by lucidrains** [comment]
@Samin100 hey Sharif, was doing some run and indeed epsilon objective converges slowly

i've added the `v-space` objective from Progressive Distillation paper (Salimans et al). you should see more immediate results

**Comment by Samin100** [comment]
@lucidrains Awesome! I'll start a new run now with it. How many steps in are you starting to see the noise change the v objective>

Also, here are a few samples from the two 32x32 training runs with the eps objective at 10k steps (batch size 1,792). 

with `latent_token_time_cond = False`
![media_images_samples_8989_53232513e8409cf79c6d](https://user-images.githubusercontent.com/13106423/222936897-23feb61d-8d7a-4019-ac71-22f5e91284ec.png)


with `latent_token_time_cond = True`
![media_images_samples_9292_67b716df531f8cbfba5a](https://user-images.githubusercontent.com/13106423/222936893-c183388f-71dc-4fac-8327-bdd6824aab8f.png)

Training loss:
![CleanShot 2023-03-04 at 17 44 55](https://user-images.githubusercontent.com/13106423/222937102-31ef0808-c974-4476-b7c5-6a1e6b35e813.png)


The samples looks more diffuse than earlier, but still no noteworthy changes yet. Gonna give the new v objective a try now.

**Comment by lucidrains** [comment]
@Samin100 yea, i agree that is strange. usually you see something much earlier, but i don't know how this perceiver-like architecture behaves with the noise objective. now i recall that the convergence i first saw was with the predict x0 objective. then some issue raised the fact that they used epsilon, and that is why i added that without testing

i think it is safe to use the predict v objective though! at least, they also used it for 'simple diffusion' with great results

**Comment by lucidrains** [comment]
using the v objective will also lend itself well to progressive distillation down the road

**Comment by lucidrains** [comment]
@Samin100 hey, thanks for opening this issue

i was digging into the ddim code this morning trying to figure out why epsilon objective samples so poorly. i think there was an issue surrounding the clipping of predicted x0 (from the predicted noise). once i rederived the predicted noise from the derived predicted x0, it seems to behave a lot better. in general, the clipping of x0 during sampling is not well documented in the literature, or even the original paper :sweat:

could you retry the predicted noise objective and see if it looks any better?

**Comment by DanielSiersleben** [comment]
Hey @lucidrains @Samin100,

I just stumbled across this issue while im trying to get RINs working for video generation with my own codebase as well.
Up to this point i wasn't able to reach anything near convergence for resolutions higher than 64x64, which i thought was possibly due to bugs but pretty much matches [this observation](https://github.com/lucidrains/recurrent-interface-network-pytorch/issues/7#issuecomment-1454954470).

But so far i only experimented with the noise prediction objective and based on your comments I'm definitely going to check out v parametrization now.

Also one thing i noticed regarding the [latent time embedding](https://github.com/lucidrains/recurrent-interface-network-pytorch/blob/b323532e40464af272a7a4e43275fb70579232ae/rin_pytorch/rin_pytorch.py#L501). Shouldn't it be `latents = latents[:, :-1]`? - But i'm still unsure about the advantage of the time embedding as latent token over the more traditional way.

But anyways thanks for open sourcing and sharing results - I will mention it if I should find out something in my following runs.

**Comment by lucidrains** [comment]
@DanielSiersleben ah shoot, yes you are correct

at a park with doggo, but let me see if I can fix this on my phone lol. welcome to the 21st century

**Comment by lucidrains** [comment]
@DanielSiersleben ok done, thank you 🙏

**Comment by lucidrains** [comment]
@DanielSiersleben i can eventually optimize the positional embeddings for video, if you let me know how your experiments turn out!

**Comment by Samin100** [comment]
@lucidrains Here are the results from a 128x128 run using the v objective with `latent_token_time_cond = False` and `dual_patchnorm = False`. Seems like it's not converging yet. I just pulled the latest changes including the fix mentioned by @DanielSiersleben and will start a new training run now and report back later today. 

Results from 128x128 model:

Start (100 steps):
![media_images_samples_101_8cbb609fc9f8246543ad](https://user-images.githubusercontent.com/13106423/222998525-5ff06d4c-cab7-4cc8-a84e-1b0dd2eae163.png)

20k steps:
![media_images_samples_19998_1e064715b0f2d5dd2bfd](https://user-images.githubusercontent.com/13106423/222998571-ae0de19d-34ba-4448-a0c2-430fab81edbe.png)

40k steps:
![media_images_samples_42117_16efadddd5640f1cacc1](https://user-images.githubusercontent.com/13106423/222998580-e59eeb94-2649-40c3-b9c1-cada7b155bfb.png)

What's interesting is that the smaller 32x32 model seemed to show signs of learning at just around 3k steps in, unlike the larger 128x128 model. All things were kept constant between the two models, except the image size and changing the `scale` argument from 1 to 0.6 for the 128px model.
![samples_3333_436aeb013610eda2fd39](https://user-images.githubusercontent.com/13106423/222998713-17478b43-1406-47a2-920a-b0b41453e0c4.png)

**Comment by lucidrains** [comment]
@Samin100 how high is your learning rate

I see results at 1k mark when using predict v objective

**Comment by lucidrains** [comment]
with the ddim patch, also seeing results much earlier, but very patchy due to the perceiver-like arch

**Comment by Samin100** [comment]
@lucidrains Are you also using 128x128 images? I'm using 2e-3 lr with the LAMB optimizer for 128x128 and adjusting accordingly from this table in the paper depending on image size. What LR are you using?

![CleanShot 2023-03-05 at 17 36 47](https://user-images.githubusercontent.com/13106423/223000473-1f553a75-ce18-4bf2-a37b-b1e609ebaf77.png)


![CleanShot 2023-03-05 at 17 35 46](https://user-images.githubusercontent.com/13106423/223000322-572679e3-929d-4bd3-8847-00c2f223d4c9.png)
It seems to converge faster than my other runs with ADAM at 1e-4. In this example the LAMB run used a larger batch size, but even with batch sizes equal LAMB is a lot faster.

**Comment by lucidrains** [comment]
![sample-1](https://user-images.githubusercontent.com/108653/223001257-979c6345-59fe-44e8-bed7-a4f1d10fc9da.png)

yup, 1k mark for predict v objective for oxford flowers

**Comment by lucidrains** [comment]
@Samin100 i [wouldn't recommend](https://arxiv.org/abs/2102.06356) lamb

**Comment by Samin100** [comment]
@lucidrains Interesting! And let me try a run with those hyperparams and ADAM right now.

**Comment by lucidrains** [comment]
![sample-2](https://user-images.githubusercontent.com/108653/223002581-eaa7c6a7-f311-4c06-821d-2fe0e47deb5e.png)

2k mark, yea it looks good

**Comment by Samin100** [comment]
It's working! I see flowers! This is 1.5k steps in:
![media_images_samples_1515_5c1fd85e2767a62e4935](https://user-images.githubusercontent.com/13106423/223017267-5ae44316-9cdd-4502-9310-c52cf3944f9b.png)

Now I'm curious why my previous runs weren't working. The biggest differences I can think of
- I was using a different dataset (Imagenet vs Oxford flowers)
- I was using fp16 with AMP, but not anymore
- LAMB optimizer instead of ADAM with a different learning rate
- the run that works is using scale = 1, whereas I was using lower vals before (0.6, 0.5)

I'll share more details as I figure things out.

**Comment by lucidrains** [comment]
@Samin100 nice! yeah would be interested to know if you can figure out the reason

**Comment by Samin100** [comment]
@lucidrains Quick update! Here's 30k steps – the model seems to be learning pretty well. I wonder how long until the seams between the image patches become seamless. 

My next step is to adjust the RIN hyperparams (one at a time) and then compare the best RIN approach to a simple diffusion approach, mostly to gain an intuition around the total compute required to train these models at varying scales.

![image](https://user-images.githubusercontent.com/13106423/223276357-7e1c1fa0-5938-41ba-bb90-a5ca026ecb4d.png)

**Comment by lucidrains** [comment]
sounds good! do reraise an issue if you find out the cause and the bug is due to the repo

**Comment by lucidrains** [comment]
@DanielSiersleben how is it looking for your video training? thumbs up or down? i'll take silence meaning that you are going heads down trying not to get scooped for your paper or startup haha, as is often the case

**Comment by DanielSiersleben** [comment]
Haha not quite, trying to get the RIN running as starting point for my Master's Thesis.

My first experiment with the v-parametrization also ran into problems but I think that was due to using AMP float16.
The gradients looked fine but I guess something with the clipping in the ddpm steps might break with AMP - but not 100% sure yet.

Running on a similar model size as you suggested, I already get promising results quite early in training on the UCF Dataset (64x64x8):
![5af6333e-0a1b-4c9a-8130-6caf2ff9e4b7](https://user-images.githubusercontent.com/35614236/223779911-15083e02-3664-46ae-91e1-6f3cf69ae3c8.gif)
Let's see where this is going. 😅 

However, with model sizes comparable to the ones used in the paper for higher resolution (256x256x8 on kinetics700), I don't seem to make any progress even after 20k steps with batch size ~64 lr 5e-4:
![906c42d5-7843-4cb3-830e-52d9ebf8cb39](https://user-images.githubusercontent.com/35614236/223781352-492f67f1-6e6e-41ac-9b63-c0793c7d73b1.gif)

Probably have to stick with smaller models...

**Comment by lucidrains** [comment]
@DanielSiersleben cool! How are you using the network? did you modify the positional embeddings at all?

If you are seeing a signal, I can spend some time and do an optimized RIN for video with some tricks I know from the lit

**Comment by Samin100** [comment]
@lucidrains  Here's the 128x128 model at 100k steps. It seems like the seams around the image patches don't want to go away. The model's hyperparams are the same ones [here](https://github.com/lucidrains/recurrent-interface-network-pytorch/issues/7#issuecomment-1455304422), except I'm using the linear noise schedule.

![image](https://user-images.githubusercontent.com/13106423/223873960-45e0721a-0892-43ac-835b-c862e490073c.png)

**Comment by lucidrains** [comment]
@Samin100 yea, you just need to train more. there's probably some architectural tweaks that can help with that artifact, but will stray from the paper

**Comment by lucidrains** [comment]
@DanielSiersleben this would also need to be a 3d conv https://github.com/lucidrains/recurrent-interface-network-pytorch/blob/main/rin_pytorch/rin_pytorch.py#L234 are you just treating the video as a grid of images?

**Comment by DanielSiersleben** [comment]
@lucidrains I mainly implemented the model on my own and was just comparing individual components with your implementation to make sure I didn't mess up.
Right now, I am keeping things pretty close to the pseudo code from the paper, so only the learnable positional embedding is added after patchifying and no PEG (which is from ViTs right?).

Tomorrow, I'll maybe add a 3D PEG and see how that works out.

What else would you modify for video besides positional embedding and patching?

**Comment by lucidrains** [comment]
nice! I think 3d peg will work well, and yes indeed it is from ViT lit

the other trick I would recommend is a token shift along the time dimension, before each feed forward block for the patches (not latents). bound to improve results

**Comment by Lamikins** [comment]
@lucidrains @Samin100 I was able to get these results at 101k steps
batch_size=128
v-pred

Results are decent but it's not nearly as compute efficient as convs (eg Karras implementation)

![media_images_sample_101000_4dfede6a67f947c3281e](https://user-images.githubusercontent.com/4073615/224626724-73740861-3f59-4d7a-ae52-793f71976429.png)

**Comment by Samin100** [comment]
@Lamikins Those results look great! Do you mind sharing the rest of the hyperparameters for this run? Specifically curious about the patch size and noise scale used. Do you think RINs might be more efficient at higher resolutions compared to a model with convs?

**Comment by ashwinipokle** [comment]
@lucidrains  I am also running into a similar issue. I am training on 32X32 CIFAR-10 images. These are the generated images after 120K steps. 

![Screen Shot 2023-03-27 at 11 19 20 PM](https://user-images.githubusercontent.com/7864463/228119318-dc160427-006b-4360-8b14-cd6787e23588.png)

I am using velocity prediction objective, Adam optimizer with learning rate of 1e-4., and batch size of 256. These are other relevant hyperparameters:

```
 model = RIN(
                dim = 128,                           
                image_size = 32,             
                patch_size = 4,              
                channels =  3, 
                dim_latent = 256,
                depth = 6,                      
                num_latents = 256,          
                latent_self_attn_depth = 4,
                latent_token_time_cond = False ,
            ).cuda()
```

Training loss curve:
![Screen Shot 2023-03-27 at 11 25 57 PM](https://user-images.githubusercontent.com/7864463/228120061-1b66c4cd-e4e6-4007-b803-b2b0c297f3fb.png)

This particular example uses sigmoid noise schedule but I have tried other noise schedules as well namely, linear, and cosine schedules without much success. Any suggestions on how to train this in a better way?

**Comment by tonyduan** [comment]
For what it's worth, I've also seen these patchy artifacts with simple datasets.

@Lamikins do you mind sharing your hyper-parameters?

**Comment by hyc9** [comment]
> @lucidrains Here's the 128x128 model at 100k steps. It seems like the seams around the image patches don't want to go away. The model's hyperparams are the same ones [here](https://github.com/lucidrains/recurrent-interface-network-pytorch/issues/7#issuecomment-1455304422), except I'm using the linear noise schedule. Do you think it's a matter of the model needing more training to "learn" better image patches?
> 
> ![image](https://user-images.githubusercontent.com/13106423/223873960-45e0721a-0892-43ac-835b-c862e490073c.png)

@Samin100  Hello, I've come across this problem and I've encountered similar issues as well. When using the RIN network structure for image reconstruction, the model's loss curve appears to have converged, but the reconstructed images exhibit block-like artifacts. I'm curious whether this is due to the inadequacy of the RIN network or the inadequacy of training. Do you have any experience with this?
